add networkd/nspawn nftables backend

fw-strlen opened the pull request:
This adds nftables/nfnetlink support to sd-netlink, then builds nftables support for firewall-util-nft backend on top of it.
See #13307  for the existing issue filed for this.

No new library dependencies are added. First patches streamline existing API by removing 'always-NULL' arguments that are not used. Then, existing firewall-util gets renamed so that it becomes possible to implement an alternative backend
to libiptc.

Next patches extend sd-netlink with a minimal nfnetlink backend.
It only supports the subset of nftables required to make masquerade/dnat work, i.e. what libiptc backend provides.

Last patch adds the nftables backend.  Iindividual commit messages have more details.

Tested:

1. test-firewall-util works
2. VM that hosts a nspawn-container will 
   add expected dnat and masquerade rules:
$ nft monitor
[ container is started now ]
add table ip io.systemd.nat
add chain ip io.systemd.nat prerouting { type nat hook prerouting priority dstnat + 1; policy accept; }
add chain ip io.systemd.nat output { type nat hook output priority -99; policy accept; }
add chain ip io.systemd.nat postrouting { type nat hook postrouting priority srcnat + 1; policy accept; }
add set ip io.systemd.nat masq_saddr { type ipv4_addr; flags interval; }
add map ip io.systemd.nat map_port_ipport { type inet_proto . inet_service : ipv4_addr . inet_service; }
add rule ip io.systemd.nat prerouting fib daddr type local dnat ip addr . port to meta l4proto . th dport map @map_port_ipport
add rule ip io.systemd.nat output ip daddr != 127.0.0.0/8 oif "lo" dnat ip addr . port to meta l4proto . th dport map @map_port_ipport
add rule ip io.systemd.nat postrouting ip saddr @masq_saddr masquerade
   new generation 2 by process 892 (systemd-network)
add element ip io.systemd.nat masq_saddr { 192.168.50.112/28 }
   new generation 3 by process 892 (systemd-network)
add element ip io.systemd.nat map_port_ipport { tcp . 2222 : 192.168.50.121 . 22 }
   new generation 4 by process 1404 (systemd-nspawn)
delete element ip io.systemd.nat map_port_ipport { tcp . 2222 : 192.168.50.121 . 22 }
[ container was shut down, then started again ]
   new generation 5 by process 1404 (systemd-nspawn)
add element ip io.systemd.nat masq_saddr { 192.168.202.240-255.255.255.255 }
    new generation 6 by process 892 (systemd-network)
add element ip io.systemd.nat map_port_ipport { tcp . 2222 : 192.168.202.249 . 22 }
    new generation 7 by process 1492 (systemd-nspawn)
delete element ip io.systemd.nat map_port_ipport { tcp . 2222 : 192.168.202.249 . 22 }
   new generation 8 by process 1492 (systemd-nspawn)
CTRL-C

3. Tested that I can connect from external machine to public-ip:2222 and get NAT'd to the container.  Same works from within the VM to container.  Using 127.0.0.1 does not work (just like with existing iptables backend).
4. Repeating same test with kernel that lacks support  for nftables dnat results in addition of iptables rules.

Caveats:

1.  ipv6 is not supported. This would be easy to add, but its not clear why one would need masquerade or dnat for ipv6 (existing iptables backend does not support it either).
 2. With nfnetlink it would be possible to detect when external entities make 'unexpected' changes to the systemd.nat table. Not implemented, as I have no idea how systemd-networkd should react to something like this. E.g. it doesn't seem like a good idea to just 'auto-undo' such changes.
3. 'set masq_saddr' doesn't handle overlaps. Example:
       fw_add_masquerade(true, AF_INET, "10.0.0.0" , 16);
       fw_add_masquerade(true, AF_INET, "10.0.0.0" , 8); /* fails */

With the iptables backend the second call works; it will just add another iptables rule.

With the nftables backend, the range 10.0.0.0-10.255.255.255 clashes with the existing range of 10.0.0.0-10.0.255.255: 2nd add gets rejected by the kernel. To resolve this it would be needed to keep track of the added elements and perform range merging when overlaps are detected. If this is needed I can add this.

4. The changeset adds two new API functions to sd-netlink to send multiple messages as one batch and to read netlink acks.
   If there is a more appropriate existing API I should use please let me know.

fw-strlen commented:
I plan to fix  the reported errors on Monday.  I'll also do a build test on a Centos7 machine to make sure it builds there as well.
I've noticed there are leftover LIBIPTC defines in some spots. Also the fw test program no longer needs to depend the libiptc build dependency.

poettering commented on src/test/test-firewall-util.c:
```diff
@@ -9,34 +9,50 @@
+        if (r == 0)
+                log_error("%s:%d Expected failure: 0 prefixlen", __FILE__, __LINE__);
+
+        errno = 0;
```
hmm? why this?

poettering commented on src/test/test-firewall-util.c:
```diff
@@ -9,34 +9,50 @@
 
         r = fw_add_masquerade(true, AF_INET, NULL, 0);
+        if (r == 0)
+                log_error("%s:%d Expected failure: NULL source", __FILE__, __LINE__);
```
please drop the `__FILE__` + `__LINE__` stuff. You get the same automatically for all log messages by setting the `SYSTEMD_LOG_LOCATION` env var to `1`. Please, let's not add redundant ways to get the sam ehere...

poettering commented on src/test/test-firewall-util.c:
```diff
@@ -9,34 +9,50 @@
 
-        r = fw_add_masquerade(true, AF_INET, NULL, 0);
+        prefixlen = 28;
+        errno = 0;
```
hmm?

poettering commented on src/shared/firewall-util.c:
```diff
@@ -1,81 +1,29 @@
 
-        return 0;
+static enum backend fw_backend_probe(void)
+{
```
our usual coding style says opening bracket for functions on the same line as the function name, see https://systemd.io/CODING_STYLE/

poettering commented on src/libsystemd/sd-netlink/netlink-types.c:
```diff
@@ -25,6 +25,9 @@
 #include <linux/wireguard.h>
 
+#include <linux/netfilter/nfnetlink.h>
+#include <linux/netfilter/nf_tables.h>
```
please keep `#include` lines sorted alphabetically, see coding style

poettering commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,323 @@
+}
+
+static int sd_nfnl_message_batch(sd_netlink *nfnl, sd_netlink_message **ret, int v)
+{
```
coding style: opening `{` on same line as function, i.e. just drop the newline here. (here and everywhere else)

poettering commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,323 @@
+
+        m = *ret;
+
+        r = sd_netlink_message_append_string(m, NFTA_CHAIN_TABLE, table);
```
we generally follow the rule that our functions should touch return values only on sucess (and then initialize them all), and on failure they shouldn't be touched at all. This function however initializes the return value initially, and then when we fail in the middle we'll leave the parameter half-initialized. 

See https://systemd.io/CODING_STYLE/#code-organization-and-semantics

fw-strlen replied:
I have convert the nfnetlink and nft plumbing accordingly.

poettering commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,323 @@
+        if (r < 0)
+                return r;
+
+        return r;
```
as above: on success initialize all return values and on failures none. 

(here and everywhere else)

poettering commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,323 @@
+
+        r = sd_netlink_message_append_data(m, NFTA_DATA_VALUE, data, dlen);
+        if (r < 0)
+              return r;
```
indentation off. please stick to multiples if 8 spaces

poettering commented on src/libsystemd/sd-netlink/netlink-socket.c:
```diff
@@ -238,6 +238,32 @@ int socket_write_message(sd_netlink *nl, sd_netlink_message *m) {
         return k;
 }
 
+int socket_writev_message(sd_netlink *nl, sd_netlink_message *m[], unsigned int msgcount) {
```
`msgcount` is ultimately the size of an object in memory hence please use `size_t` type

poettering commented on src/libsystemd/sd-netlink/netlink-socket.c:
```diff
@@ -238,6 +238,32 @@ int socket_write_message(sd_netlink *nl, sd_netlink_message *m) {
 
+int socket_writev_message(sd_netlink *nl, sd_netlink_message *m[], unsigned int msgcount) {
+        struct iovec *iovs;
+        unsigned int i;
```
as above size_t

poettering commented on src/libsystemd/sd-netlink/netlink-socket.c:
```diff
@@ -238,6 +238,32 @@ int socket_write_message(sd_netlink *nl, sd_netlink_message *m) {
+        }
+
+        k = writev(nl->fd, iovs, msgcount);
+        free(iovs);
```
consider using `_cleanup_free_`instead

poettering commented on src/libsystemd/sd-netlink/netlink-socket.c:
```diff
@@ -238,6 +238,32 @@ int socket_write_message(sd_netlink *nl, sd_netlink_message *m) {
+
+        for (i = 0; i < msgcount; i++) {
+                assert(m[i]->hdr);
+                assert(m[i]->hdr->nlmsg_len);
```
we usually try to avoid using C's downgrade-to-bool feature for stuff that is neither pointer nor boolean/boolean-like. i.e. for number checks we usually spell out `!= 0`, so that the reader groks that this is a number, not a boolean.

poettering commented on src/libsystemd/sd-netlink/sd-netlink.c:
```diff
@@ -227,6 +227,31 @@ int sd_netlink_send(sd_netlink *nl,
+                assert_return(!messages[i]->sealed, -EPERM);
+                rtnl_seal_message(nl, messages[i]);
+                if (serial)
+                        serial[i] = rtnl_message_get_serial(messages[i]);
```
as above, do not clobber return values on failure

fw-strlen replied:
I've resolved this by moving the allocation of serial[ ] from the caller to this function. Caller now passes "uint32_t **serials"
instead of "*serial". **serial is only updated on success.

poettering commented on src/libsystemd/sd-netlink/sd-netlink.c:
```diff
@@ -227,6 +227,31 @@ int sd_netlink_send(sd_netlink *nl,
 
+int sd_netlink_sendv(sd_netlink *nl,
+                     sd_netlink_message *messages[],
+                     unsigned int msgcount,
```
size_t

poettering commented on src/libsystemd/sd-netlink/sd-netlink.c:
```diff
@@ -227,6 +227,31 @@ int sd_netlink_send(sd_netlink *nl,
+                     sd_netlink_message *messages[],
+                     unsigned int msgcount,
+                     uint32_t *serial) {
+        unsigned int i;
```
size_t

poettering commented on src/libsystemd/sd-netlink/sd-netlink.c:
```diff
@@ -614,21 +614,15 @@ int sd_netlink_call_async(
         return k;
```
btw, doesn't matter, but we do not use Signed-off-by in systemd that's a kernel thing

fw-strlen replied:
Yes, matter of habit. I dropped it.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+#include "macro.h"
+#include "socket-util.h"
+
+#include <stdio.h>
```
please keep public headers always before private ones

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+
+static const char nft_systemd_dnat_map_name[] = "map_port_ipport";
+static const char nft_systemd_table_name[] = "io.systemd.nat";
+static sd_netlink *nfnl;
```
global static variables is something we really don't like, because of threading and so on. Please find a different place to store this, maybe pass in a generic context object or so to these functions, where you can store this

fw-strlen replied:
What about something like this:
src/shared/firewall-util.h:

typedef struct fw_util_src ctx fw_util_ctx;
int fw_util_ctx_new(fw_util_ctx **);
 int fw_add_masquerade(
+               fw_util_ctx *ctx,
                 bool add,
There would be a firewall-util-private.h with
struct fw_util_ctx {
       union {
               /* backend specific .... */

The nft backend would then keep the nfnetlink handle in this union.
The object would have to be managed by nspawn/networkd.

I think if this route is taken it would make sense to open a new pull request that only has the fw-util API rework without
any of the nftables changes. What do you think? Any better idea?

fw-strlen replied:
I've added a fw_ctx that gets passed around; it now contains what was formerly kept in 'static' variable. Please clarify/advise if thats not what you had in mind.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+static const char nft_systemd_table_name[] = "io.systemd.nat";
+static sd_netlink *nfnl;
+
+#define NFNL_DEFAULT_TIMEOUT 1000000ULL
```
what time unit is this? please suffix the define name with `_USEC` or so, to make that clear. please also consider using USEC_PER_SEC and similar defines, to make this easier to grok when looking at it

fw-strlen replied:
done.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+#include <stdio.h>
+
+static const char nft_systemd_dnat_map_name[] = "map_port_ipport";
+static const char nft_systemd_table_name[] = "io.systemd.nat";
```
hmm, static strings like this are a bit weird, why not just use a `#define` here? code should result in the same

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+
+static int nfnl_netlink_sendv(sd_netlink *nl,
+                              sd_netlink_message *messages[],
+                              unsigned int msgcount) {
```
`size_t` here and everywhere else where we count memory or objects in memory

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+
+        serial = new0(uint32_t, msgcount);
+        if (!serial)
+               return -1;
```
please do not make up error messages like this. always return errors from the `errno` vocabulary, i.e. `return -ENOMEM`

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+               /* If message is an error, this returns embedded errno */
+               r = sd_netlink_read(nfnl, serial[i], NFNL_DEFAULT_TIMEOUT, &ret);
+               if (r < 0)
+                     return r;
```
6ch indenting? please stick to multiples of 8 spaces!

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+               if (r < 0)
+                     return r;
+
+               sd_netlink_message_unref(ret);
```
hmm, you read the message and drop it immediately?

i think it would better to change the sd_netlink_read() function so that it only optionally returns the message it read, i.e. by passing NULL to it, so that the function can free it internally

fw-strlen replied:
Actually it already provides that feature. I've fixed the caller to pass NULL.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        return sd_netlink_message_close_container(m); /* NFTA_LIST_ELEM */
+}
+
+static int __nfnl_add_expr_lookup_set(sd_netlink_message *m, const char *set_name,
```
hmm, double udnerscores are actually reserved by the compiler/C language. kernel uses them anyway, but let's avoid that here, and use single underscores at most at most.

That said, can't we find a better name for this, i.e. using human understandable words to discern it from nfnl_add_expr_lookup_set() that lacks the underscores?

fw-strlen replied:
Yes, ugly habit of mine. I've renamed the functions and dropped the "__" prefix.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                 return r;
+
+        sd_netlink_message_close_container(m); /* NFTA_RULE_EXPRESSIONS */
```
error checking missing

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+/* this allows nft to format the map_port_ipport concatenated set */
+static uint32_t concat_types2(enum nft_key_types a, enum nft_key_types b)
+{
+        return a << 6 | b;
```
maybe add an assert here that b doesn't grow above 2^6-1 and a not above 2^24-1

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        return a << 6 | b;
+}
+
+#define NFT_INIT_MSGS	16
```
where does this come from?

fw-strlen replied:
Its a consequence of my choice to pass the netlink messages in the same transactions as pointer-to-array, so that
array needs to be large enough to hold all netlink message buffers.

An alternative solution might be to use use existing 'next' pointer in sd_netlink_message and chain all of the
messages (add sd_netlink_message_append(sd_netlink *a, sd_netlink *b)?)  socket_write_message() could be changed to build an iovec based on the number of chained multipart messages.  Would ease freeing as sd_netlink_message_unref() already follows the chain.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+}
+
+#define NFT_INIT_MSGS	16
+static int __fw_nftables_init(sd_netlink_message *ruleset[], unsigned int msgcnt)
```
double underscores...

poettering replied:
please spell out the size of the array then in the parameter list (and add a comment to the `#define`) :

```c
#define NFTABLE_RULESET_MESSAGES_MAX 16 /* We know we need 16 entries at most */
static int fw_nftables_generate_ruleset(
                sd_netlink_message ruleset[static NFT_RULESET_MESSAGES_MAX], 
                size_t msgcnt) {
…
```

i.e. the "static" stuff in array params is a helpful hint to compilers (primarily llvm right now) to know what size to expect.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        int r;
+
+        /* Set F_EXCL so table add fails if the table already exists. */
+        r = sd_nfnl_nft_message_new_table(nfnl, &ruleset[++msgcnt], nft_systemd_table_name, NLM_F_EXCL | NLM_F_ACK);
```
if this fails, then msgcnt will be left increased but the object not actually increased. I am pretty sure the counter should only be increased after the function returned successfully

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
```
hmm, assert_return() we use for APIs that are user facing, i.e. that either already are export to be linked to by other programs that use libsystemd, or are written so that they eventually are. sd-netlink is one of those libs (which is supposed to be exposed to clients eventually, but currently is still private). For stuff that is not user-facing, but facing only to our own code we just use plain assert(). Both shall be used only for programming errors (and not runtime errors), the only difference is that we are a bit stricter for ourselves then for users, hence if we encounter a programming error in our own code we abort, but if the user made one we try to go on.

fw-strlen replied:
That makes sense. These are all "programmer error" asserts. I've replaced it with plain assert().

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+
+int fw_nftables_init(void)
+{
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnlp;
```
will result in memory corruption if we exit early from this function. needs to be initialized to NULL so that the automatic destructor becomes a NOP when notthing is allocated to it yet

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (nfnl)
+                return 0;
+
+        memset(ruleset, 0, sizeof(ruleset));
```
initialize at time of declaration? i.e. `= {}`

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                return r;
+
+        nfnl = sd_netlink_ref(nfnlp);
```
hmm, what? you ref twice?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+                                           unsigned int prefixlen)
+{
+        union in_addr_union source_end = *source;
+        unsigned int nplen = 32 - prefixlen;
```
assert() first that prefixlen <= 32?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+{
+        union in_addr_union source_end = *source;
+        unsigned int nplen = 32 - prefixlen;
+        uint32_t v = (1 << nplen);
```
assert first that nplen <= 31?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+
+        v = (1 << nplen);
+
+        source_end.in.s_addr += htonl(v);
```
please do not use htonl() in our codebase. it's really weird, since it suggests it would return "long" but actually returns uint32_t. Also, "network byte order" is just legacy talk for "big endian". Hence, please htobe32() instead, which is much clearer.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (!source || source_prefixlen == 0)
+               return -EOPNOTSUPP;
+
+        memset(transaction, 0, sizeof(transaction));
```
initialize when your decalre? `= {}`

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        } else {
+                 r = sd_nfnl_nft_message_del_setelems_begin(nfnl, &transaction[1], nft_systemd_table_name, "masq_saddr");
+                 if (r < 0)
+                        return r;
```
do the `if (r < 0) return r` check outside of the `if (add)` branches? it's going to be a bit shorter then!

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                return r;
+
+        return 0;
```
the allocated messages are not freed here?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        assert(add || !previous_remote);
+
+        if (af != AF_INET)
+                return -EOPNOTSUPP;
```
maybe EAFNOSUPPORT is a better error here?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+                return -EOPNOTSUPP;
+
+        if (!IN_SET(protocol, IPPROTO_TCP, IPPROTO_UDP))
+                return -EOPNOTSUPP;
```
EPROTONOSUPPORT?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+
+        data[1] = htobe16(remote_port);
+
+        memset(transaction, 0, sizeof(transaction));
```
initialize at moment of declaration?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                return r;
+
+        return 0;
```
the messages are leaked?

fw-strlen replied:
I will run the existing test case through valgrind to make sure this is resolved correctly.

fw-strlen replied:
Indeed, I've fixed this. valgrind --leak-check=full ./build/test-firewall-util
now reports "All heap blocks were freed".

poettering commented on src/shared/firewall-util.c:
```diff
@@ -13,12 +13,18 @@ enum backend {
+        int ret = fw_nftables_init();
+
+        if (ret == 0)
+               return FW_BACKEND_NFTABLES;
```
why declare `int ret`? You could as well just write `if (fw_nftables_init() == 0)` here...

poettering reviewed (changes requested):
would love a test case for this (even if it's a manual one, given that the CIs might not have nft), but i guess we had none before...

poettering commented:
regarding the overlap issue: if people set things up that way things are broken a bit, no? maybe leave it that way hence, but propagate a recognizable error up (EUCLEAN?) and make networkd/nspawn print out a friendly explanation about it

poettering commented:
would love ipv6 support for this. for the nspawn case I am pretty sure it makes a ton of sense to hide the fact that an nspawn container is used in the back...

poettering commented:
regarding multiple messages in netlink support, i figure @yuwata knows this better by now than me...

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,853 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
```
This commit msg should probably contain "Fixes: #13307" somewhere, to link up the commit with the issue in a way github understands

poettering commented:
BTW, the ipv6 support is being asked for in #10254, but be excellent to close this at the same time

poettering commented:
Hmm, does this address #6106 too? if not, maybe it should?

fw-strlen commented:
> Hmm, does this address #6106 too? if not, maybe it should?

I think that should be done in an extra request. I'm also not sure this should be changed, the behaviour is there for some time.
e.g. what if someone has a service listening on 127.0.0.1:12345 and there is a port forward for that from external:12345 to internal:42 ? If that is resolved then after update connect to 127.0.0.1:12345 ends up in the container.

fw-strlen commented:
> BTW, the ipv6 support is being asked for in #10254, but be excellent to close this at the same time

Unless there are extreme objections, I would prefer to do this in a different PR.  HOWEVER, given that reply i would make sure that the next iteration would already pass the appropriate information (esp. the inet family) in function arguments internally so that code churn in sd-netlink is minimized. Does that sound reasonable?

poettering commented:
> I think that should be done in an extra request. I'm also not sure this should be changed, the behaviour is there for some time.
> e.g. what if someone has a service listening on 127.0.0.1:12345 and there is a port forward for that from external:12345 to internal:42 ? If that is resolved then after update connect to 127.0.0.1:12345 ends up in the container.

That's pretty unlikely, no? Might be relevant enough to mention in NEWS, but maybe too unlikely to make any bigger fuss about? Dunno?

> Unless there are extreme objections, I would prefer to do this in a different PR. HOWEVER, given that reply i would make sure that the next iteration would already pass the appropriate information (esp. the inet family) in function arguments internally so that code churn in sd-netlink is minimized. Does that sound reasonable?

Doing that in a separate PR is entirely OK. I am thankful for every issue closed, but it's entirely up to you if you want to do that, and when, and in how many PRs!

yuwata commented:
Not reviewed yet. But please rebase.

fw-strlen commented:
> Not reviewed yet. But please rebase.

I've rebased and pushed my current working tree. Its till not 100% complete, I will work on this tomorrow.
I've added two more commits, one commit to add a 'fw context' to avoid the need to keep a static nfnetlink handle in the nft backend and another one to rework the way ip_tables module gets loaded.  The other commits have been changed mostly to address bugs and code format fixups, they do not have major changes. I will push another version+comment once i think all reported problems have been resolved.

fw-strlen commented:
> > BTW, the ipv6 support is being asked for in #10254, but be excellent to close this at the same time
> 
> Unless there are extreme objections, I would prefer to do this in a different PR. HOWEVER, given that reply i would make sure that the next iteration would already pass the appropriate information (esp. the inet family) in function arguments internally so that code churn in sd-netlink is minimized. Does that sound reasonable?

FWIW I have a (small) patch that adds ipv6 support.  I would make another PR after this one is merged if thats OK.
Main issue is that it doesn't do anything yet -- at this time all callers only pass ipv4 addresses.
Do you have any preference on how to change networkd and nspawn? I.e., should there be new options to enable ipv6 explicitly ("IPV6Masquerade=true") or similar? No hurry, but something to keep in mind.

fw-strlen commented:
> regarding the overlap issue: if people set things up that way things are broken a bit, no? maybe leave it that way hence, but propagate a recognizable error up (EUCLEAN?) and make networkd/nspawn print out a friendly explanation about it

In case this would happen networkd would trigger this error message (already in master):
        r = address_establish(address, link);
        if (r < 0)
                log_link_warning_errno(link, r, "Could not enable IP masquerading, ignoring: %m");
(errno will be EEXIST).

fw-strlen commented:
I think you can risk a second look at this pr (head is 4eaf4db, "src: remove remaining libiptc entanglemen"). Feedback welcome, will try to address further change requests in a timely fashion.

poettering commented:
> "IPV6Masquerade=true

In various cases we have "per-proto bools", i.e. where Masquerade=no means off, =yes means on on both ipv4 and ipv6, and =ipv4 means ipv4-only an =ipv6 means ipv6-only.

poettering commented on src/shared/firewall-util.h:
```diff
@@ -6,13 +6,20 @@
 
+typedef struct fw_ctx fw_ctx;
+
+int fw_ctx_new(fw_ctx **);
```
please give return params names (and ideally the name `ret` here, which is what we typically do for the return parameter if there's only one)

poettering commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,310 @@
+        sd_netlink_message *m;
+        int r;
+
+        r = nft_message_new(nfnl, ret, family, NFT_MSG_NEWCHAIN, NLM_F_CREATE | NLM_F_ACK);
```
you initialize the return param here already even though later code in this function can fail, and leave the function half-done returning a half-setup object. The usual pattern to follow is this:

```
_cleanup(object_freep) ObjectType *o = NULL;
int r;

o = allocate_object();
if (!o)
        return -ENOMEM;

r  = object_initialize_something(o);
if (r < 0)
        return r;

r  = object_initialize_something_else(o);
if (r < 0)
        return r;

*ret = TAKE_PTR(o);
return 0;
```

i.e. allocate object into temporary, internal variable (here: `o`), set it up fully, and after everything went successfully assign it to the return param, not a moment before. And if we exit the function on some error path early, then the `_cleanup_` logic will make sure the object is freed again and the caller will never know because guarantee that on error paths the return params are untouched while on success they are unconditionally initialized.

poettering replied:
(this applies the same to all functions below the same: please never touch return params until everything succeeded and cannot fail anymore)

fw-strlen replied:
I'm sorry, this was result of a broken squash commit, i.e. a later commit changed it to the expected pattern but the initial added commit still contained the half-initialized version. I've now amened/squashed the correct change, the line
above now works with "m" and assigns "ret" only until right before 0 is returned.

poettering commented on src/libsystemd/sd-netlink/sd-netlink.c:
```diff
@@ -227,6 +227,41 @@ int sd_netlink_send(sd_netlink *nl,
+int sd_netlink_sendv(sd_netlink *nl,
+                     sd_netlink_message *messages[],
+                     size_t msgcount,
+                     uint32_t **serial) {
```
we usually prefix return params `ret_`, please do so here too, i.e. `serial`→ `ret_serial` to make it easier to see what is output.

(if there's only a single return param and and it's the primary purpose of the function to return it, you may also call the return param just `ret`, i.e without any further suffix)

poettering commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -96,15 +97,13 @@ int sd_nfnl_nft_message_new_basechain(sd_netlink *nfnl, sd_netlink_message **ret
                 return r;
 
-        m = *ret;
-
```
this looks like a broken squahs: should have been squashed into the earlier commit that added the function?

fw-strlen replied:
Yes, exactly. I've fixed the squash so that all of the added code has the same pattern from the start, i.e.:
_cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
...
*ret = TAKE_PTR(m);
return 0.

poettering commented on src/shared/firewall-util-private.h:
```diff
@@ -5,18 +5,45 @@
+                struct {
+                        sd_netlink *nfnl;
+                } nft;
+        } u;
```
i'd not bother with the subunion and substruct... It#s not that a million backends are going to pop-up any time soon which want to overload this...

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,906 @@
+                              sd_netlink_message *messages[],
+                              size_t msgcount) {
+        _cleanup_free_ uint32_t *serial = NULL;
+        unsigned int i;
```
doesnt't matter, but we usually use just "unsigned" rather than "unsigned int"

poettering replied:
that said, this should actually be size_t I guess, given that you just count through `msgcount` elements and that param is rightfully size_t

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,906 @@
+void fw_nftables_exit(fw_ctx *ctx) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnl = NULL;
+
+        nfnl = TAKE_PTR(ctx->u.nft.nfnl);
```
```c
ctx->u.ft.nfnl = sd_netlink_unref(ctx->u.ft.nfnl);
```

fw-strlen replied:
changed to ctx->nfnl = sd_netlink_unref(ctx->nfnl); -- i dropped the embedded union/struct as per your other suggestion.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,906 @@
+        if (r < 0)
+                return r;
+
+        return r;
```
return 0 i guess otherwise the if check a few lines above makes little sense

fw-strlen replied:
indeed. changed.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,906 @@
+        r = nfnl_netlink_sendv(nfnl, transaction, tsize);
+
+out_unref:
+        while (tsize)
```
we usually avoid relying on C's downgrade-to-bool feature for numerics, but use explicit comparisons for them, to not create the impression something was used as a bool, while it is actually a number. i.e. `while (tsize > 0)…` would be better

fw-strlen replied:
sorry about this, changed to while (tsize > 0) here and elsewhere.

poettering commented on src/shared/firewall-util-iptables.c:
```diff
@@ -20,15 +21,45 @@
+
+        if (r == 0) {
+                const char *cmdline[] = {
+                       "modprobe",
```
uh, please do not use modprobe... use libkmod if anything, but shelling out is ugly

I must say I am not much of a fan of this... most of our services run without CAP_SYS_MODULE, and this would change here...

fw-strlen replied:
Okay, i've dropped this change.

poettering commented on src/core/kmod-setup.c:
```diff
@@ -85,10 +85,6 @@ int kmod_setup(void) {
-#if HAVE_LIBIPTC
-                /* netfilter is needed by networkd, nspawn among others, and cannot be autoloaded */
-                { "ip_tables", "/proc/net/ip_tables_names", false,  false,   NULL      },
-#endif
```
distros that keep this enabled shouldn't mind about this auto-loading, no?

fw-strlen replied:
Ok, fair enough.  I've tossed the commit that dropped the ip_tables autoload from kmod-setup. I've retained the snipped that enables the --port option unconditionally and merged that part into the preceeding change.

fw-strlen commented on src/nspawn/nspawn.c:
```diff
@@ -184,6 +184,7 @@ static char *arg_image = NULL;
 static char *arg_oci_bundle = NULL;
 static VolatileMode arg_volatile_mode = VOLATILE_NO;
 static ExposePort *arg_expose_ports = NULL;
+static fw_ctx *firewall_ctx = NULL;
```
Renamed the type to FirewallContext.

fw-strlen commented on src/network/networkd.c:
```diff
@@ -112,6 +113,10 @@ static int run(int argc, char *argv[]) {
 
+        r = fw_ctx_new(&m->firewall_ctx);
+        if (r < 0)
+                log_warning_errno(r, "Could not init firewall, IPMasquerade option not available: %m");
```
done.

fw-strlen commented on src/shared/firewall-util.c:
```diff
@@ -95,13 +91,14 @@ int fw_add_masquerade(
         size_t sz;
         struct nf_nat_ipv4_multi_range_compat *mr;
         int r;
+        int protocol = 0;
```
done.

fw-strlen commented on src/shared/firewall-util.c:
```diff
@@ -1,81 +1,30 @@
-                entry->ip.dst = destination->in;
-                in4_addr_prefixlen_to_netmask(&entry->ip.dmsk, destination_prefixlen);
-        }
+static enum backend fw_backend;
```
Renamed the enum type and variable name to FirewallBackend.

fw-strlen commented on src/shared/firewall-util.c:
```diff
@@ -1,81 +1,30 @@
 
-        return 0;
+static enum backend fw_backend_probe(void)
+{
```
Ugh. My bad. I've corrected this now an will make another push once build test completes.

fw-strlen commented on src/shared/firewall-util.c:
```diff
@@ -27,17 +17,39 @@ static enum backend fw_backend_probe(void)
+         return 0;
+}
+
+void fw_ctx_free(fw_ctx *ctx) {
```
I've change it accordingly and changed the callsites to assign (null) the freed object.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,902 @@
+                              sd_netlink_message *messages[],
+                              size_t msgcount) {
+        _cleanup_free_ uint32_t *serial = NULL;
+        unsigned int i;
```
still: size_t

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,902 @@
+                        r = tmp;
+        }
+
+        return r < 0 ? r : 0;
```
just `return r;` should be equivalent no?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,902 @@
+                return r;
+
+        proto_reg = NFT_REG32_02;
+        r = nfnl_add_expr_dnat(m, family, NFT_REG32_01, proto_reg);
```
you could just specify that directly, no? via the indirect operatio via the proto_reg variable?

fw-strlen replied:
Yes. Once ipv6 support gets added this needs a minor tweak however, because the ipv6 address will occupy
registers 01..04, i.e. protocol will be stored in REG_05 for ipv6. I would leave it as is, its also a bit more descriptive
this way, i think.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,902 @@
+                return r;
+
+        /* 1st statement (cont.): bitops/prefix */
+        r= nfnl_add_expr_bitwise(m, NFT_REG32_01, NFT_REG32_01, &lomask, &zero, sizeof(lomask));
```
missing space after r

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,902 @@
+        for (i = 0; i < NFT_INIT_MSGS; i++) {
+                if (!batch[i])
+                        break;
+                sd_netlink_message_unref(batch[i]);
```
the non-NULL check of batch[i] should not be necessary, as sd_netlink_message_unref() should be a NOP anyway if NULL is pased

fw-strlen replied:
yes, removed.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,902 @@
+                r = 0;
+
+out_unref:
+        for (i = 0; i < NFT_INIT_MSGS; i++) {
```
doesn't matter but you could also iterate to `msgcnt` here, instead of NFT_INIT_MSGS, and it would be a tiny bit quicker...

fw-strlen replied:
done.

fw-strlen commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,328 @@
+        int r = sd_nfnl_message_add_setelem(m, num, key, klen, data, dlen);
+        if (r < 0)
+                return r;
+        return r;
```
I've renamed the static helper called here to the exported name and removed this code.

fw-strlen commented on src/network/networkd.c:
```diff
@@ -112,6 +113,10 @@ static int run(int argc, char *argv[]) {
 
+        r = fw_ctx_new(&m->fw_ctx);
+        if (r < 0)
+                log_warning_errno(r, "Could not initialize firewall, IPMasquerade= option not available: %m");
```
Can you clarify this a bit? I can change the api like this:
+int fw_add_masquerade(
+                FirewallContext **ctx,
+                bool add,
+                int af,
+                const union in_addr_union *source,
+                unsigned int source_prefixlen);
+int fw_nftables_add_local_dnat(
+                FirewallContext **ctx,
[..]
(Note '**ctx').
Both functions would check *ctx == NULL and call fw_ctx_new() [which would be an internal helper].
Question is what to do on failure. I think it would make sense to still return the initialized context even on error,
which violates coding guidelines. Otherwise, each future call would try to init/probe the kernel again for the missing feature again, as we can't distinguish NULL (first call) from 'already tried, no working backend available'.
For nspawn this might not be a big deal when the error is changed to be fatal.
But for networkd I'd think exiting if masquerade support isn't available is kinda broken. Any suggestion?
If you think that the extra alloc/probe/free isn't that much of an issue I can change this as outlined above.

fw-strlen commented on src/nspawn/nspawn.c:
```diff
@@ -186,6 +186,7 @@ static char *arg_image = NULL;
 static char *arg_oci_bundle = NULL;
 static VolatileMode arg_volatile_mode = VOLATILE_NO;
 static ExposePort *arg_expose_ports = NULL;
+static FirewallContext *fw_ctx = NULL;
```
I will wait here until the 'do we want explicit initialization or not' clears up. If implicit/lazy is the way to go then different context could be used as well, as the end of run only has to 'clean up exposed ports', i.e. this only occurs once. Doing an extra
context init for that doesn't look like a problem and it would avoid having to pass the context from run up the call chain.

yuwata commented:
Could you also rebase this again? Sorry, I made a large refactoring for networkd.

fw-strlen commented:
I've pushed a new version for review, but please do not merge it, the topmost commit will tell you why.
I've altered the API towards 'lazy' init (which would allow to remove fw_ctx_new()).  If you prefer this way, let me know and I will do a new push with those changes squashed.

mbiebl commented:
Is there a chance to get this merged before v247 is released?

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,899 @@
+        if (r < 0)
+                return r;
+
+        if (flags) {
```
doesn't matter much, but we try to avoid relying on C's downgrade-to-bool feature except for pointers, and stuff that is boolean or boolean-like. i.e. please change to `if (flags != 0)`.

but doesn't really matter much

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,899 @@
+};
+
+static uint32_t concat_types2(enum nft_key_types a, enum nft_key_types b) {
+        return a << TYPE_BITS | b;
```
hmm, i think it would be safer to cast `a` to `(uint32_t)` here first, to ensure that the value is wide enough for the `<<` to not overflow. C compilers have a lot of freedom how to map enums to numeric types after all.

poettering commented on src/shared/firewall-util.c:
```diff
@@ -8,7 +8,10 @@
-static enum FirewallBackend FirewallBackendProbe(void) {
+static enum FirewallBackend FirewallBackendProbe(FirewallContext *ctx) {
+        if (fw_nftables_init(ctx) == 0)
+               return FW_BACKEND_NFTABLES;
```
should we cache the result?

fw-strlen replied:
The caller stores the return value in the FirewallContext, the probe function is only called once per context.

poettering commented on src/shared/firewall-util.c:
```diff
@@ -48,15 +48,20 @@ FirewallContext *fw_ctx_free(FirewallContext *ctx) {
-        if (!ctx)
-                return -ENOTCONN;
+        if (!*fw_ctx) {
+                int ret = fw_ctx_new(fw_ctx);
```
doesn't matter, but we almost always call the variable we collect errors in `int r` and declare it at the outermost function scope. might be something to follow here too, just keep surprise minimal

poettering reviewed (changes requested):
Looks really good. Just some smaller fixes.

poettering commented:
I cherry-picked the first two commits into git master already btw. They are obviously good, and something that clearly looked like good to merge ahead of time.

poettering commented:
also needs rebase

fw-strlen commented on src/nspawn/nspawn.c:
```diff
@@ -4859,11 +4865,16 @@ static int run_container(
+                struct add_change_args expose_args = {
+                        .exposed = *exposed,
+                        .fw_ctx = fw_ctx,
+                };
```
Doh. Probably "worked" due to inlining. I've moved the struct to the run function so it has same storage duration
as the in_addr_union.

fw-strlen commented on src/libsystemd/sd-netlink/nfnl-message.c:
```diff
@@ -0,0 +1,318 @@
+
+        r = sd_netlink_message_open_container(m, NFTA_CHAIN_HOOK);
+        if (r < 0)
+                goto cancel;
```
Yes, right. fixed.

fw-strlen commented:
Current branch head (1e8dc84a043ddf1) takes care of the feedback/review comments (thanks!).

I've tested:
- systemd on kernel with no nftables uses iptables backend
- systemd on recent kernel uses nft backend
- container run via machinectl (and configured/exposed service) results in expected entry in the mapping table
- stopping container removes the entry
- masquerade for the dhcp range was added as well

bluca commented:
Needs another rebase

fw-strlen commented:
Rebased, was only a minor context clash with 88fc9c9bad006a9fd05fbe4fe0107a2adccdae3c.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,904 @@
+        return sd_netlink_message_close_container(m);
+}
+
+
```
doesn't matter, but there are two empty lines here

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,904 @@
+        assert(prefixlen <= 32);
+        nplen = 32 - prefixlen;
+
+        mask = (1U << nplen) - 1U;
```
if npen is allowed to be 32 (i.e. prefixlen is allowed to be zero), then this will result in undefined behaviour, since 1U << 32 is not defined, and will overflow.

if prefixlen is now allowed to be zero, then maybe error out early?

(it looks to me as if a full ip range should be ok to support here though, hence maybe special case prefixlen == 0 here and allow it?)

fw-strlen replied:
Its not allowed anymore, this will yield -EINVAL earlier.
The reason is two-fold:
1. The caller is networkd, and afaics it always passes in the netmask configured on the interface.
2. The existing iptables backend (which the nft one is modeled on) doesn't restrict based on interface. IOW, when using /0, it will masquerade everything on any interface, which is probably not wanted.  If this should be supported, it might be best to add extra special code that just deletes the existing rule (which checks if a given address is in a range we're masquerading for) and replaces that with a single masquerade rule with no further checks.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -0,0 +1,904 @@
+
+out_unref:
+        while (tsize > 0)
+                sd_netlink_message_unref(transaction[--tsize]);
```
in similar cases we added foobar_unref_many() wrappers, which take two params: the array and the size, and it goes through all entries and unrefs them. in this file you already have this loop twice, might be worth adding here too

fw-strlen replied:
I was reluctant to add/expose a new sd-netlink api for this, but i can surely add it if thats wanted.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -779,6 +779,22 @@ static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+        return 0;
+}
+
+
```
double empty lines...

fw-strlen replied:
removed.

poettering commented on src/shared/firewall-util-nft.c:
```diff
@@ -779,6 +779,22 @@ static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+        }
+
+        return 0;
+}
```
i think the commit msgs should probably be comments here in the sources...

poettering commented:
looks great, only minor stuff left that we can ignore – except for the netmask overflow thing which looks to me as if it is relevant and should be fixed.

fw-strlen commented:
current incarnation (bc5a9b82d5e8649dcd753cd5f2a90eeb07526563) has following changes compared to last reviewed one:
1. rebase on master (no conflicts)
2. remove two blank lines
3. add commit message of topmost commit also as comment as per Lennarts suggestion

I did not (yet?) add the new _unreg_many helper and did not special case (allow) 0 prefixlen so far.

poettering commented:
let's merge this, the rest we can figure out and tweak after the merge.

excellent work! thank you!

DaanDeMeyer commented:
> I think that should be done in an extra request. I'm also not sure this should be changed, the behaviour is there for some time.
e.g. what if someone has a service listening on 127.0.0.1:12345 and there is a port forward for that from external:12345 to internal:42 ? If that is resolved then after update connect to 127.0.0.1:12345 ends up in the container.

@fw-strlen If I wanted to make the necessary changes to make this work, what changes would be required? I tried removing the `ip daddr != 127.0.0.1/8` rule from the output chain but that's not enough it seems. I then tried to enable `net.ipv4.conf.ve-image.route_localnet=1` for the container's veth link host interface but that didn't do it either. I did go from a hanging ssh command without disabling route_localnet to getting `ssh: connect to host localhost port 8729: No route to host` when disabling it.
